### PR TITLE
BUG: GDCMImageIO accepted images with non-orthogonal direction cosine.

### PR DESCRIPTION
**Pull request by zivy**
The DICOM standard requires that "The row and column direction cosine
vectors shall be orthogonal, i.e., their dot product shall be
zero. The row and column direction cosine vectors shall be normal,
i.e., the dot product of each direction cosine vector with itself
shall be unity."

ITK images support arbitrary affine matrices as the direction
cosine. When they are not compliant with the DICOM requirements GDCM
write silently changes the direction cosine to be compliant. This
commit enforces the requirement in ITK so the user can choose if they want to
explicitly modify the matrix to comply with the DICOM requirements or
possibly write using a different format which supports affine matrices
for direction cosine.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/IO/GDCM/test/itkGDCMImageOrientationPatientTest.cxx [hunk 1]
```diff
@@ -103,16 +103,43 @@ itkGDCMImageOrientationPatientTest(int argc, char * argv[])
...
+    direction3D[2][1] = .6;
+    direction3D[0][2] = 0;
+    direction3D[1][2] = 1;
+    direction3D[2][2] = 0;
+    value.str("");
+    value << direction3D[0][0] << "\\" << direction3D[1][0] << "\\" << direction3D[2][0] << "\\" << direction3D[0][1]
+          << "\\" << direction3D[1][1] << "\\" << direction3D[2][1];
+    itk::EncapsulateMetaData<std::string>(dictionary, "0020|0037", value.str());
+    src2D->GetOutput()->SetMetaDataDictionary(dictionary);
+    try
```
Comment (dzenanz) [inline Modules/IO/GDCM/test/itkGDCMImageOrientationPatientTest.cxx:133]: Definitely use [ITK_TRY_EXPECT_NO_EXCEPTION](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3rc03/Modules/Core/TestKernel/include/itkTestingMacros.h#L115).

File: Modules/IO/GDCM/test/itkGDCMImageOrientationPatientTest.cxx [hunk 2]
```diff
@@ -103,16 +103,43 @@ itkGDCMImageOrientationPatientTest(int argc, char * argv[])
   writer2D->SetInput(src2D->GetOutput());
   writer2D->SetFileName(filename.str().c_str());
 
+  // Cause intentional error, non-orthogonal direction cosine,
+  // write should fail.
   try
```
Comment (dzenanz) [inline Modules/IO/GDCM/test/itkGDCMImageOrientationPatientTest.cxx:108]: Maybe use [ITK_TRY_EXPECT_EXCEPTION](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3rc03/Modules/Core/TestKernel/include/itkTestingMacros.h#L98)?

**Review by dzenanz — COMMENTED** [review]
Generally looks good.

File: Modules/IO/GDCM/test/itkGDCMImageOrientationPatientTest.cxx (hunk 2 above)
Comment (zivy) [reply Modules/IO/GDCM/test/itkGDCMImageOrientationPatientTest.cxx]: done

**Review by dzenanz — APPROVED** [review]

File: Modules/IO/GDCM/src/itkGDCMImageIO.cxx [hunk 3]
```diff
@@ -1104,6 +1105,11 @@ GDCMImageIO::Write(const void * buffer)
       image.SetDirectionCosines(5, 0);
     }
   }
+  gdcm::DirectionCosines gdcmDirection(image.GetDirectionCosines());
+  if (!gdcmDirection.IsValid())
+  {
+    itkExceptionMacro("Invalid direction cosine, non-orthogonal or unit length.");
```
[the remark this answers is not in the document]
Comment (zivy) [reply Modules/IO/GDCM/src/itkGDCMImageIO.cxx]: done.

**Review by dzenanz — APPROVED** [review]

**Pull request merged**